File: implement-shell-tools/cat/cat.mjs
import { program } from "commander";
import { promises as fs } from "node:fs";

program.name("cat");
program.description("simple cat clone");
program.option("-n, --number", "number all output lines");
program.option("-b, --number-nonblank", "number non-empty output lines");
program.argument("[paths...]", "file(s) to process");
program.parse();

const { number, numberNonblank } = program.opts();
const paths = program.args;

async function processFile(path) {
  try {
    const stat = await fs.stat(path);

    if (stat.isDirectory()) {
      console.error(`${path}: Is a directory`);
      return;
    }

    const content = await fs.readFile(path, { encoding: "utf-8" });
    const lines = content.split("\n");
    const hasTrailingNewLine = content.endsWith("\n");
    const effectiveLines = hasTrailingNewLine ? lines.slice(0, -1) : lines;
    let lineNumber = 1;

    const processed = effectiveLines.map((line) => {
      if (numberNonblank) {
        if (line.trim() !== "") {
          const result = `${lineNumber}\t${line}`;
          lineNumber += 1;
          return result;
        }

        return line;
      }

      if (number) {
        const result = `${lineNumber}\t${line}`;
        lineNumber += 1;
        return result;
      }

      return line;
    });

    const output = processed.join("\n");
    console.log(output);
  } catch (error) {
    console.error(`${path}: ${error.message}`);
  }
}

try {
  for (const path of paths) {
    await processFile(path);
  }
} catch (error) {
  console.error(error.message);
}

File: implement-shell-tools/cat/cat.py
import argparse
import os

parser = argparse.ArgumentParser(prog="cat",
                                 description="simple cat clone")

parser.add_argument("-n",
                    "--number",
                    dest="number_all",
                    action="store_true",
                    help="number all output lines")

parser.add_argument("-b",
                    "--number-nonblank",
                    dest="number_nonblank",
                    action="store_true",
                    help="number non-empty output lines")

parser.add_argument("paths", nargs="+", help="file(s) to process")

args = parser.parse_args()

def process_line(line_number, line):
    if args.number_nonblank:
        if line == "\n":
            print(line, end="")
            return line_number
        
        print(f"{line_number:6}\t{line}", end="")
        return line_number + 1
    
    if args.number_all:
        print(f"{line_number:6}\t{line}", end="")
        return line_number + 1
    
    print(line, end="")
    return line_number

for path in args.paths:
    if os.path.isdir(path):
        print(f"cat: {path}: Is a directory")
        continue

    with open(path, "r") as file:
        line_number = 1
        
        for line in file:
            line_number = process_line(line_number, line)
